bot/exts/utils/attachment_pastebin_uploader.py

```diff
@@ -0,0 +1,144 @@
+
+import re
+
+import aiohttp
+import discord
+from discord.ext import commands
+from pydis_core.utils import paste_service
+
+from bot.bot import Bot
+from bot.constants import Emojis
+from bot.log import get_logger
+
+log = get_logger(__name__)
+
+PASTEBIN_UPLOAD_EMOJI = Emojis.check_mark
+DELETE_PASTE_EMOJI = Emojis.trashcan
+
+
+class EmbedFileHandler(commands.Cog):
+    """
+    Handles automatic uploading of attachments to the paste bin.
+
+    Whenever a user uploads one or more attachments that is text-based (py, txt, csv, etc.), this cog offers to upload
+    all the attachments to the paste bin automatically. The steps are as follows:
+    - The bot replies to the message containing the attachments, asking the user to react with a checkmark to consent
+        to having the content uploaded.
+    - If consent is given, the bot uploads the contents and edits its own message to contain the link.
+    - The bot DMs the user the delete link for the paste.
+    - The bot waits for the user to react with a trashcan emoji, in which case the bot deletes the paste and its own
+        message.
+    """
+
+    def __init__(self, bot: Bot):
+        self.bot = bot
+        self.pending_messages = set[int]()
+
+    @staticmethod
+    async def _convert_attachment(attachment: discord.Attachment) -> paste_service.PasteFile:
+        """Converts an attachment to a PasteFile, according to the attachment's file encoding."""
+        encoding = re.search(r"charset=(\S+)", attachment.content_type).group(1)
+        file_content = (await attachment.read()).decode(encoding)
+        return paste_service.PasteFile(content=file_content, name=attachment.filename)
+
+    @commands.Cog.listener()
+    async def on_message_delete(self, message: discord.Message) -> None:
+        """Allows us to know which messages with attachments have been deleted."""
+        self.pending_messages.discard(message.id)
+
+    @commands.Cog.listener()
+    async def on_message(self, message: discord.Message) -> None:
+        """Listens for messages containing attachments and offers to upload them to the pastebin."""
+        # Check if the message contains an embedded file and is not sent by a bot.
+        if message.author.bot or not any(a.content_type.startswith("text") for a in message.attachments):
+            return
+
+        log.trace(f"Offering to upload attachments for {message.author} in {message.channel}, message {message.id}")
+        self.pending_messages.add(message.id)
+
+        # Offer to upload the attachments and wait for the user's reaction.
+        bot_reply = await message.reply(
+            f"Please react with {PASTEBIN_UPLOAD_EMOJI} to upload your file(s) to our "
+            f"[paste bin](<https://paste.pythondiscord.com/>), which is more accessible for some users."
+        )
+        await bot_reply.add_reaction(PASTEBIN_UPLOAD_EMOJI)
+
+        def wait_for_upload_permission(reaction: discord.Reaction, user: discord.User) -> bool:
```

Comment:
I don't like this inline predicate midway through the function, if we need to use variables from a higher scope then a method specifically to create the predicate at the top of the file is much cleaner.